app/models/ppm_estimator.rb

```diff
@@ -0,0 +1,216 @@
+  end
+
+  def dest_svc_area
+    @dest_svc_area ||= ServiceArea.from_date_and_svc_area_number(effective_400ng_date, end_zip3.service_area)
+  end
+
+  def distance
+    @distance ||= DtodZip3Distance.dist_mi(start_zip3.zip3, end_zip3.zip3).to_i
+  end
+
+  def effective_400ng_date
+    # If the user enters a date that's before or after all 400NG rates in the database, then
+    # use the 400NG rates with an effective date closest to the date the user gave
+    # the FullUnpacks table is the shortest, so should be the quickest to query
+    @effective_400ng_date ||=
+      if (effective_range = FullUnpack.effective_date_range).cover?(date)
+        date
+      elsif date < effective_range.min
+        effective_range.min
+      else
+        effective_range.max
+      end
+  end
+
+  def effective_tdl_date
+    # If the user enters a date that's before or after the TDLs in the database, then
+    # use the discounts with a TDL closest to the date the user gave
+    @effective_tdl_date ||=
+      if (tdl_range = TopTspByChannelLinehaulDiscount.tdl_date_range).cover?(date)
+        date
+      elsif date < tdl_range.min
+        tdl_range.min
+      else
+        tdl_range.max
+      end
+  end
+
+  def rate_area(zip3, zip5)
+    # you can usually get the rate_area from just the ZIP3
+    area = zip3.rate_area
+    # Sometimes, all 5 digits of the ZIP code are needed to get the rate area
+    return area unless area == 'ZIP'
+    Zip5RateArea.select(:rate_area).find_by(zip5: zip5).rate_area
+  end
+
+  def inv_linehaul_discount
+    @inv_linehaul_discount ||= TopTspByChannelLinehaulDiscount.inv_discount(
+      # the origin is a rate area
+      rate_area(start_zip3, estimator_params[:start].to_i),
+      # For CONUS moves, destination is the region.
+      # If the orig and dest are in the same state, the region is 15!
+      start_zip3.state == end_zip3.state ? 15 : end_zip3.region,
+      effective_tdl_date
+    )
+  end
+
+  def base_rate
+    return BaseLinehaul.rate(effective_400ng_date, distance, full_weight) unless full_weight < 1000
+    # pro-rate the 1000 lb baseline rate for shipments less than 1000 lbs
+    BaseLinehaul.rate(effective_400ng_date, distance, 1000) * (full_weight / 1000.0)
+  end
+
+  def linehaul_charges
+    base_rate + (orig_svc_area.linehaul_factor + dest_svc_area.linehaul_factor) * cwt + shorthaul
+  end
+
+  def shorthaul
+    return 0 if distance > 800
+    Shorthaul.rate(effective_400ng_date, cwt, distance)
+  end
+
+  def non_linehaul_charges
+    full_pack_cost + cwt * (orig_svc_area.orig_dest_service_charge + dest_svc_area.orig_dest_service_charge + FullUnpack.rate(effective_400ng_date, dest_svc_area.services_schedule))
+  end
+
+  def discount_range
+    # don't go below 0% or above 100% before applying PPM incentive %
+    @discount_range ||= ([inv_linehaul_discount - 0.02, 0].max * 0.95)..([inv_linehaul_discount + 0.02, 1].min * 0.95)
+  end
+
+  class << self
```

Comment:
I'm not superfan of `class << self` for defining a block of class methods. In my limited experience, it's been easy to lose track of the methods that fall within this block based on indentation alone. Seeing methods named `def self.foo` is much easier.

Opinions vary, though, and it's a balance of terseness vs. legibility, so… use your best judgment.

Reply:
I've decided to just make these methods instance methods. Ruby is kludgy with what I would call "static" methods, and there was no end-user benefit.